filebeat/input/filestream/prospector.go

```diff
@@ -129,9 +131,82 @@ func (p *fileProspector) Init(
 	//  - path
 	if identifierName != fingerprintName {
 		p.logger.Debugf("file identity is '%s', will not migrate registry", identifierName)
+	} else {
+		p.logger.Debug("trying to migrate file identity to fingerprint")
+		prospectorStore.UpdateIdentifiers(func(v loginp.Value) (string, interface{}) {
+			var fm fileMeta
+			err := v.UnpackCursorMeta(&fm)
+			if err != nil {
+				return "", nil
+			}
+
+			fd, ok := files[fm.Source]
+			if !ok {
+				return "", fm
+			}
+
+			// Return early (do nothing) if:
+			//  - The identifiers are the same
+			//  - The old identifier is neither native nor path
+			oldIdentifierName := fm.IdentifierName
+			if oldIdentifierName == identifierName ||
+				!(oldIdentifierName == nativeName || oldIdentifierName == pathName) {
+				return "", nil
+			}
+
+			// Our current file (source) is in the registry, now we need to ensure
+			// this registry entry (resource) actually refers to our file. Sources
+			// are identified by path. However, as log files rotate the same path
+			// can point to a different file.
+			//
+			// So, to ensure we're dealing with the resource from our current file,
+			// we use the old identifier to generate a registry key for the current
+			// file we're trying to migrate, if this key matches with the key in the
+			// registry, then we proceed to update the registry.
+			registryKey := v.Key()
+			oldIdentifier, ok := identifiersMap[oldIdentifierName]
+			if !ok {
+				// This should never happen, but we properly handle it just in case.
+				// If we cannot find the identifier, move on to the next entry
+				// some identifiers cannot be migrated
+				p.logger.Errorf(
+					"old file identity '%s' not found while migrating entry to "+
+						"new file identity '%s'. If the file still exists, it will be re-ingested",
+					oldIdentifierName,
+					identifierName,
+				)
+				return "", nil
+			}
+			previousIdentifierKey := newID(oldIdentifier.GetSource(
+				loginp.FSEvent{
+					NewPath:    fm.Source,
+					Descriptor: fd,
+				}))
+
+			// If the registry key and the key generated by the old identifier
+			// do not match, log it at debug level and do nothing.
```

Comment:
Where is the log?

Reply:
This piece of code has not been touched by this PR (Git is confused, `prospectorStore.TakeOver` is added, it is not a modification of existing code): https://github.com/elastic/beats/blob/9e29bcb1044ce37bf6d13b00784a1fa5f79f0540/filebeat/input/filestream/prospector.go#L184-L188

Given the length of this PR I rather not make unrelated changes. What do you think?